### PR TITLE
Implement automatic SSH tunneling for HPG

### PR DESCRIPTION
**Pull request by fdr**
Implement automatic SSH tunneling for HPG
When Bastion config vars are present, use them to open a tunnel to
rewrite the URI of the database to the tunnel.

Do this for `pg:psql`, commands like `pg:kill` (which are similar
non-interactive variants), and `pg:push` and `pg:pull`.  Because the
tunneling requires a process to stay alive, switch from "exec"ing psql
to "spawn"ing it as a subprocess, but *only* in the case where
tunneling is required as a hedge against change regular users using
un-tunneled psql.

Example use of pg:psql, on an app that has the bastion-requiring
database "DATABASE" and the regular database "COPPER":

    $ heroku config
    DATABASE_BASTIONS:             @ref:blooming-slyly-8913:bastions
    DATABASE_BASTION_KEY:          @ref:blooming-slyly-8913:bastion-key
    DATABASE_BASTION_REKEYS_AFTER: @ref:blooming-slyly-8913:bastion-rekeys-after
    DATABASE_URL:                  @ref:blooming-slyly-8913:url
    HEROKU_POSTGRESQL_COPPER_URL:  @ref:reading-surely-4913:url

    $ heroku pg:psql -c "select 'hello world'"
    ---> Connecting to DATABASE_URL
      ?column?
    -------------
     hello world
    (1 row)

    $ heroku pg:psql COPPER -c "select 'hello world'"
    ---> Connecting to HEROKU_POSTGRESQL_COPPER_URL
      ?column?
    -------------
     hello world
    (1 row)

Because of the patch to `#exec_sql`, one-shot commands like `pg:kill`,
`pg:outliers`, etc work too:

    $ heroku pg:kill 1234
    WARNING:  PID 1234 is not a PostgreSQL server process
     pg_cancel_backend
    -------------------
     f
    (1 row)

    $ heroku pg:table-size
     name | size
    ------+------
    (0 rows)

Finally, pg:push and pg:pull work (as apparent to the user) exactly
the same as before:

    $ heroku pg:pull DATABASE test
    pg_dump: reading schemas
    pg_dump: reading user-defined tables
    pg_dump: reading extensions
    [...]

    $ heroku pg:push test DATABASE
    [...]

### CODE REVIEW

File: lib/heroku/command/pg.rb [hunk 1]
```diff
@@ -726,19 +736,6 @@ def wait_for(attach, interval)
     end
   end
 
-  def find_uri
-    return @uri if defined? @uri
-
-    attachment =  generate_resolver.resolve(shift_argument, "DATABASE_URL")
-    if attachment.kind_of? Array
```
Comment (fdr) [inline lib/heroku/command/pg.rb]: I don't really understand this type switch very well. It seems incongruous with other cases with how generate_resolver.resolve is used, so is it dead code?

**Comment by will** [comment]
This patch broke pg:diagnose

```
heroku ➤ git bisect run bin/heroku pg:diagnose IVORY -a customerapp
running bin/heroku pg:diagnose IVORY -a customerapp
 !    Unknown database. Valid options are: HEROKU_POSTGRESQL_IVORY_URL, HEROKU_POSTGRESQL_PINK_URL
Bisecting: 49 revisions left to test after this (roughly 6 steps)
[0312712ec2940dd32a104bbe5124aac56b08d9f2] Merge pull request #1691 from heroku/space-apps-unjoined
running bin/heroku pg:diagnose IVORY -a customerapp
 !    Unknown database. Valid options are: HEROKU_POSTGRESQL_IVORY_URL, HEROKU_POSTGRESQL_PINK_URL
Bisecting: 24 revisions left to test after this (roughly 5 steps)
[d0d9f7b4ab46aa5bfe5c1852ccacb5cb8fe98745] v3.40.8
running bin/heroku pg:diagnose IVORY -a customerapp
 !    Unknown database. Valid options are: HEROKU_POSTGRESQL_IVORY_URL, HEROKU_POSTGRESQL_PINK_URL
Bisecting: 12 revisions left to test after this (roughly 4 steps)
[a1ddda6b6aedae58913af0a280ce5e3b4926e794] Update RELEASE.md
running bin/heroku pg:diagnose IVORY -a customerapp
 !    Unknown database. Valid options are: HEROKU_POSTGRESQL_IVORY_URL, HEROKU_POSTGRESQL_PINK_URL
Bisecting: 5 revisions left to test after this (roughly 3 steps)
[fbae53fb3c3c371d952f371585d4da0b0e4eac17] v3.40.7.pre
running bin/heroku pg:diagnose IVORY -a customerapp
 !    Unknown database. Valid options are: HEROKU_POSTGRESQL_IVORY_URL, HEROKU_POSTGRESQL_PINK_URL
Bisecting: 3 revisions left to test after this (roughly 2 steps)
[b13a751b89ac123970b87d8d266bebb7c76ccfed] Merge pull request #1673 from heroku/fix-account-verification
running bin/heroku pg:diagnose IVORY -a customerapp
Report 2ade09ef-5cdf-48f2-a138-c54f01b60c8d for customerapp::HEROKU_POSTGRESQL_IVORY_URL
available for one month after creation on 2015-10-13 20:32:16.63562+00

RED: Hit Rate
Name                    Ratio
----------------------  -------------------
overall index hit rate  0.8423341115274596
overall cache hit rate  0.8939453088883605

YELLOW: Indexes
Reason                  Index                                                       Index Scan Pct  Scans Per Write  Index Size  Table Size
----------------------  ----------------------------------------------------------  --------------  ---------------  ----------  ----------
Never Used Indexes      public.tablename::somesortof_index                          0.00            0.00             118 MB      1493 MB

YELLOW: Load
  Load 1.195

GREEN: Connection Count
GREEN: Long Queries
GREEN: Idle in Transaction
GREEN: Bloat
GREEN: Blocking Queries
Bisecting: 1 revision left to test after this (roughly 1 step)
[c08f2ed40e109ac64efa9319e8f02c47daff4d7d] Merge pull request #1670 from fdr/pgsql-ssh
running bin/heroku pg:diagnose IVORY -a customerapp
 !    Unknown database. Valid options are: HEROKU_POSTGRESQL_IVORY_URL, HEROKU_POSTGRESQL_PINK_URL
Bisecting: 0 revisions left to test after this (roughly 0 steps)
[194cfd5245c1e9ce60dde111d09d0a22676cc042] Implement automatic SSH tunneling for HPG
running bin/heroku pg:diagnose IVORY -a customerapp
 !    Unknown database. Valid options are: HEROKU_POSTGRESQL_IVORY_URL, HEROKU_POSTGRESQL_PINK_URL
194cfd5245c1e9ce60dde111d09d0a22676cc042 is the first bad commit
commit 194cfd5245c1e9ce60dde111d09d0a22676cc042
Author: Daniel Farina <daniel@heroku.com>
Date:   Mon Jul 27 21:48:41 2015 -0700

    Implement automatic SSH tunneling for HPG

    When Bastion config vars are present, use them to open a tunnel to
    rewrite the URI of the database to the tunnel.

    Do this for `pg:psql`, commands like `pg:kill` (which are similar
    non-interactive variants), and `pg:push` and `pg:pull`.  Because the
    tunneling requires a process to stay alive, switch from "exec"ing psql
    to "spawn"ing it as a subprocess, but *only* in the case where
    tunneling is required as a hedge against change regular users using
    un-tunneled psql.

    Example use of pg:psql, on an app that has the bastion-requiring
    database "DATABASE" and the regular database "COPPER":

        $ heroku config
        DATABASE_BASTIONS:             @ref:blooming-slyly-8913:bastions
        DATABASE_BASTION_KEY:          @ref:blooming-slyly-8913:bastion-key
        DATABASE_BASTION_REKEYS_AFTER: @ref:blooming-slyly-8913:bastion-rekeys-after
        DATABASE_URL:                  @ref:blooming-slyly-8913:url
        HEROKU_POSTGRESQL_COPPER_URL:  @ref:reading-surely-4913:url

        $ heroku pg:psql -c "select 'hello world'"
        ---> Connecting to DATABASE_URL
          ?column?
        -------------
         hello world
        (1 row)

        $ heroku pg:psql COPPER -c "select 'hello world'"
        ---> Connecting to HEROKU_POSTGRESQL_COPPER_URL
          ?column?
        -------------
         hello world
        (1 row)

    Because of the patch to `#exec_sql`, one-shot commands like `pg:kill`,
    `pg:outliers`, etc work too:

        $ heroku pg:kill 1234
        WARNING:  PID 1234 is not a PostgreSQL server process
         pg_cancel_backend
        -------------------
         f
        (1 row)

        $ heroku pg:table-size
         name | size
        ------+------
        (0 rows)

    Finally, pg:push and pg:pull work (as apparent to the user) exactly
    the same as before:

        $ heroku pg:pull DATABASE test
        pg_dump: reading schemas
        pg_dump: reading user-defined tables
        pg_dump: reading extensions
        [...]

        $ heroku pg:push test DATABASE
        [...]

:100644 100644 4e70129e8746608f5aaac92cc84546e2079bd0e7 56f868e31f33c6453c789932747a6146ba3d687a M  Gemfile.lock
:100644 100644 c5a258f6a3b1f373b5985260cd78bdf3c6d7e593 38bfbb02504965dc0f57ae9700fef20ca8cdab4b M  heroku.gemspec
:040000 040000 5d381427d8d5f3c6d9a506b684a38ee87f2c1e72 e87c0ee36ebaaffdee58c4c871ca8b9b70da4002 M  lib
:040000 040000 59b6c11a9a1e5c92e19765862c0181372baa4c57 8acc1987582e943dd84fc54bc17c678bb70efdd0 M  spec
bisect run success
```

**Comment by fdr** [comment]
On Tue, Oct 13, 2015 at 1:35 PM Will Leinweber <notifications@github.com>
wrote:

> This patch broke pg:diagnose
>

Heh. Does it break it in every case, or shorthands only?

**Comment by will** [comment]
Every case. The version check relied on  
```
def find_uri
  return @uri if defined? @uri
```
so it didn't resolve twice.

**Comment by fdr** [comment]
Wait, is it fixed now? I cannot reproduce using `bundle exec bin/heroku` on the very recent commit c38f111ffa6c5139f19f341d13978d317c51a41f.

**Comment by fdr** [comment]
I also cannot reproduce on the pre-pg:diagnose-change 8d805f04bf49e8294f138817f6e69d381f7ed75f.

**Comment by will** [comment]
Yes, it's fixed now. I mentioned you [in the PR that fixed it](https://github.com/heroku/heroku/pull/1775)

**Comment by fdr** [comment]
Hm. Do you think I should grovel around for more similar defects, or do you suspect that things are now more fine-as-is?

The motivation behind my changes you mention in that PR is that, to allow use of Bastions, it's necessary to treat a database attachment as a cohort of config vars: just the `URL` part does not suffice like it would before.

When the cohort has that bastion information, then I taught some of the primitives in the system to handle tunneling transparently.

So part of this patch was teaching (apparently not quite) all the relevant call sites how to deal with a richer, multi-config var "attachment" value rather than a URI, and it would appear that in the shuffle I lost a caching behavior that was meaningful.

**Comment by will** [comment]
That looks to be the only use of `@uri` so everything else is probably fine

**Pull request merged**